### ISSUE TITLE
Stereoscopy with Vulkan?

### ISSUE THREAD

**Issue by Geenz** [ConfettiFX/The-Forge#144]
Just wondering if this is anywhere on the roadmap.  As it stands this doesn't seem to work with a fairly standard 120Hz display and shutter glasses.

**Comment by wolfgangfengel** [comment]
I am not sure I understand what you are asking for? What is your usage case? We used this for VR apps in the past.

**Comment by Geenz** [comment]
In this case, we are *not* utilizing a VR headset.  Think more along the lines of a CAVE application where you may have one or more displays running at 120Hz and a pair of 3D shutter glasses that's "flipping" between each frame in sync with the display.  I've tried doing this just by having two cameras (left and right) that effectively "flip" between their view mats.  This doesn't appear to work however.

**Comment by wolfgangfengel** [comment]
We have people who worked on the CAVE and a bunch of VR apps, let me ask them ... I get back.

**Comment by dsrour** [comment]
@Geenz Would you mind sharing some snippets of code of what you are doing?  Is there a way we can reproduce the issues you are seeing on our side without your hardware?

For a CAVE like environment, you usually would have multiple processes driving each display with synchronization being done via some sort of IPC mechanism over a LAN.

**Comment by Geenz** [comment]
@dsrour I think you misunderstand.

I'm looking for stereoscopic rendering here, not running this across multiple machines.  The latter is simple enough.

I literally just need stereoscopy as in shutter glasses here, similar to NVIDIA's 3D Vision.  My understanding is to achieve this you basically setup a swapchain with two layers, and setup commands accordingly.

**Comment by manaskulkarni786** [comment]
Do you mean VK_KHR_multiview?

**Comment by Geenz** [comment]
In essence, yes.

**Comment by manaskulkarni786** [comment]
It's not yet supported. We will add it soon.

**Comment by Geenz** [comment]
Is there a rough timeline for this?  I imagine it'd be low priority for you guys.

**Comment by wolfgangfengel** [comment]
We will look into it.

**Comment by Geenz** [comment]
Given that I needed it this week, I went ahead and rolled an implementation into my company's fork of The Forge.  I can look into pushing it upstream if there's any interest.

**Comment by mgsim** [comment]
I, for one, would be very interested. I am also interested in using stereoscopy with the Forge. I'm still waiting on best practices. See issue #101.

**Comment by Geenz** [comment]
@mgsim I've pushed support to my fork at https://github.com/Geenz/The-Forge.  Note that you'll need to create your shaders in accordance with multiview's spec - in otherwords make sure to enable GL_KHR_multiview in your shader, then access any properties via gl_ViewIndex.  You'll want to pass in a view mask, and optionally a correlation mask to bindRenderTargets - definitely not ideal, but seems to do the trick.  In the future it'll probably be a good idea to have a `BindRenderTargetDesc` or some such to define stuff like this.

Also, make sure you modify your `SwapChainDesc` to have `mEnableStereo` set to true.

I'll submit a PR once I've gotten DX12 up and running with stereo.  Unfortunately, there isn't an exact 1:1 correlation between Vulkan's multiview and DX12 texture arrays - and stereo is handled a bit different in DX12 in general.  However selection by texture array slice in DX12 shouldn't be hard for rendering to texture arrays.

**Comment by mgsim** [comment]
@Geenz Thank you so much. I'll have a look at your fork.

**Comment by wolfgangfengel** [comment]
We are looking into this.

**Comment by Geenz** [comment]
Hey everyone!

I haven't checked in on TF in a while - I was wondering what the status of this was.

**Comment by wolfgangfengel** [comment]
Hi, we have Quest support now in the public release. I think that should close this issue?

**Comment by Geenz** [comment]
@wolfgangfengel It does add stereo for VR (using multiview as well which is nice) - though it'd be nice to see more generalized VR support in this regard.  Looking at the code, I don't see much that _wouldn't_ work with desktop VR headsets beyond the fact some bits have been ifdef'd out for Quest support specifically, so something to consider.  I have a project that requires fairly robust VR support that I've been messing with Forge for a while now with, so maybe I can contribute something upstream here?

When I originally opened this issue, I was building CAVE systems - so a VR headset wasn't exactly something we were targetting at the time.  Though general multiview/view instanced rendering would also be nice given you can squeeze a little bit more performance out of some operations on some hardware - like rendering out directional light shadow cascades.
